### PR TITLE
[WIP - Do not merge] Add VOF interface tracking plugin

### PR DESCRIPTION
**Pull request by class4kayaker**
This is an implementation of extremely basic functionality for a 2D VoF based
passive interface tracking algorithm plugin. There are a few locations where
what is likely to be an undesirable approach has been used to allow coupling of
the VOF model to ASPECT while avoiding a non-trivial modification to the
main code during the initial stages.

Current features:
- Signed distance Level Set initialization
- Basic interface error estimation calculation for testing
- Passive VOF interface tracker w/ basic vtu output

Limitations:
- Only 2D has any implementation
- Only parallelogram meshes will function correctly
- Canot handle variation in refinement level
- Dimensionally split update

Hacks:
- Stores VoF data in duplicate DoFHandler and Triangulation
  - If means to add new state data of non-standard type (DGP(0)) added, fix
    should be trivial
- Relies on index consistency for relation between Triangulations
  - Only relevant while two separate DoF Handlers are necessary

Planned Improvements:
- Fix limitations
- Plugin based output (not necessary if VoF data moved to main DoFHandler)
- Reduce computation to narrow band around interface
- Try coupled LS VoF method

### CODE REVIEW

**Comment by bangerth** [comment]
Wow - massive amount of work here! The patch is marked as work in progress but let us know if you want feedback. In that case it would also help if you pointed out places you'd like someone to look at first, given the size of it all.

**Comment by class4kayaker** [comment]
I would definitely appreciate any feedback you have to offer.

The first part that I would like some feedback on would be the use of a copy of the Triangulation and a separate DoFHandler. At least personally, this does not feel like the best implementation as it would require code overhead to ensure that the plugin's copy remains up to date with that maintained by the main code. There is currently no means by which to include a non-standard separate state variable into the main set of state variables, so I used the separate DoFHandler as a workaround. While this currently works I do feel that a better approach would be to maintain the volume fraction data in the same data structure as the rest of the state variables unless there is a reason not to.

**Comment by bangerth** [comment]
Let's try to morph the patch over time and incrementally, rather than trying to address everything in one step.

So let's talk about the Triangulation/DoFHandler thing first. There is really no need to copy the Triangulation. But then I also don't actually see that you do -- my interpretation is that you just store a pointer to the one you get from the `SimulatorAccess` object, right? There, instead of storing a pointer to it (that may or may not be current any more), the usual style is to simply ask the `SimulatorAccess` base class for a reference to the triangulation whenever you need it. That's a change I would definitely make.

For the `DoFHandler`: Since relatively recently (see issue #772), there is a way for other parts of the code to declare variables such as the ones you need for your work. Can you take a look and see whether you can use this mechanism? This may avoid the need for a separate DoFHandler variable in your plugin.

File: source/vofinterface/QCMid.cc [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  double h = 1.0 / n;
+
+  if (n == 0)
+    return;
+
+  for (i = 0; i < n; ++i)
+    {
+      this->quadrature_points[i] = Point<1> ((i+0.5) * h);
+      this->weights[i] = h;
+    }
```
Comment (bangerth) [inline source/vofinterface/QCMid.cc]: What you have here is actually the iterated midpoint rule. You can already get that via `QIterated<dim> q(QMidpoint<1>(), n);`. Is there a good reason to duplicate this functionality? My suggestion would be to simply use the facilities already present in deal.II -- i.e., to remove this class and file, and use the way to write the same thing stated above.

File: source/vofinterface/QCTrap.cc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+  this->quadrature_points[0] = Point<1> (0.0);
+  this->weights[0] = 0.5 * b_weight;
+  for (i = 1; i < n; ++i)
+    {
+      this->quadrature_points[i] = Point<1> (i * b_weight);
+      this->weights[i] = b_weight;
+    }
+  this->quadrature_points[n] = Point<1> (1.0);
+  this->weights[n] = 0.5 * b_weight;
```
Comment (bangerth) [inline source/vofinterface/QCTrap.cc]: The same is true -- what you have here is equivalent to `QIterated<dim> (QTrapez<1>(), n);`.

File: source/vofinterface/QCMid.cc (hunk 1 above)
Comment (class4kayaker) [inline source/vofinterface/QCMid.cc]: I must have not seen the Iterated option or misunderstood what it was before looking into it, and this is almost certainly an undesirable duplication. I am currently making sure that I did not overlook a difference, but you are almost certainly correct.

**Comment by class4kayaker** [comment]
I currently have to store an entirely separate copy of the triangulation to avoid issues with the order in which things get released by aspect causing exceptions. While it doesn't cause any interruptions during the code, without the copy of the triangulation it would consistently produce erroneous exceptions that might cover other issues.

I will look into the new signal for additional state variables. It looks like it will be possible to use it to make the necessary modifications, though I may need to figure out how to adjust my update code to continue to be correct. To confirm, it will be possible to ensure that there is no advection applied to the new variable, as the VOF field that needs to be stored should not be updated using a standard FEM advection step. Looking at the Signals. it looks like the set_assemblers signal might be what I am looking for, though if so, I am not certain how to establish the initial data set for the declared component variable.

Given the likely desirability of reducing the area required for calculation to a restricted region around the interface, if there is no reasonable means to restrict the calculated area. An alternate, possibly desirable approach might be to continue to maintain a separate Triangulation and DoFHandler. This would introduce a communication and maintenance overhead

**Comment by bangerth** [comment]
> I currently have to store an entirely separate copy of the triangulation
> to avoid issues with the order in which things get released by aspect
> causing exceptions. While it doesn't cause any interruptions during the
> code, without the copy of the triangulation it would consistently
> produce erroneous exceptions that might cover other issues.

I don't, obviously, understand all of the intricacies of this, but 
duplicating the triangulation would seem the wrong solution for the 
problem. I would rather want to try and understand what errors you see 
when using the existing triangulation, and how we can address these.


> I will look into the new signal for additional state variables. It looks
> like it will be possible to use it to make the necessary modifications,
> though I may need to figure out how to adjust my update code to continue
> to be correct. To confirm, it will be possible to ensure that there is
> no advection applied to the new variable, as the VOF field that needs to
> be stored should not be updated using a standard FEM advection step.
> Looking at the Signals. it looks like the set_assemblers signal might be
> what I am looking for, though if so, I am not certain how to establish
> the initial data set for the declared component variable.

Yes, set_assemblers() is the way you can achieve what you want. I think 
you should be able to achieve initialization by duplicating the way this 
is done for the compositional fields, for example, just for your 
specific field.


> Given the likely desirability of reducing the area required for
> calculation to a restricted region around the interface, if there is no
> reasonable means to restrict the calculated area. An alternate, possibly
> desirable approach might be to continue to maintain a separate
> Triangulation and DoFHandler. This would introduce a communication and
> maintenance overhead

This does not seem to be a path to happiness, if you ask me :-) You 
would have to duplicate all of the work we have already done to ensure 
parallel scalability, the ability to checkpoint/restart everything, etc. 
Try to hook into the functionality we already have.

In trying to understand the patch, I realize that I do not actually 
understand what volume-of-fluid really is, and what the goal of all of 
this is. Can you summarize what the method is and what you want to use 
it for?

**Comment by class4kayaker** [comment]
The goal of the patch is to implement a volume-of-fluid base interface tracking method. That is, the desired result is to implicitly track a region of fluid. This is done by reconstructing the interface every timestep using volume fraction information on the cells and the assumption that at any single point cna only have one type of fluid, then using the reconstructed interface to obtain the fluxes. In general, it is useful for some of the same problems as a level set interface tracking method, but is better about volume conservation, as can be expected given that it is formulated around the volume conservation equation. The most general use is probably to track a region of fluid which may have some relevant properties. As compared to a particle approach to tracking fluid volume, a volume of fluid method would have a far lower overhead due to a very significant reduction in the amount of calculation necessary for a certain resolution.

I agree that that duplication of the triangulation, etc is likely to not be a desirable approach. The only reason I even considered it was to attempt to find some benefit to maintaining a separate DoFHandler if it had been necessary. If the VoF method was not coupled to aspect, there might be a significant calculation cost benefit to restricting the calculation to a small region around the interface. Given that this is being coupled to aspect, though, it is likely not worth the overhead.

**Comment by class4kayaker** [comment]
I have continued to look over the signals approach, and I am currently unsure how to design my code such that it would not need to be maintained as a separate library, but would not always be active, especially considering that the current algorithm includes some required quantities being calculated and stored. Also, the current implementation requires a dimensionally split approach, which might make it more reasonable to delay some of this until the unsplit algorithm is done.

**Comment by class4kayaker** [comment]
After a bit more inspection, I believe that I have a better understanding of how the signals are designed to work, and I can get the . I am currently working on building up the implementation using the signals approach as a separate branch, so as to allow a more gradual modification to the code. I think that the best way to handle this might be to open a new pull request for said branch but leave this one open until that branch has reached a comparable level of feature completeness. If there is a more preferable convention, I will use that.

I currently think that I will probably require (and will implement) at least one new signal for initialization. An initial state of the interface is required but I do not currently see a signal that would allow this to be handled. This signal might also be useful for other improvements such as the particle initialization, so it may make sense to implement that separately from this patch.

Also, examining the signal sets suggests that the set_assemblers signal does not have an appropriate assembler for handling assembly and solutions of additional variables, though I may be misunderstanding the design. As such, my current impression is that a separate assemble/solve would be needed, though due to the nature of the VoF method(effectively a finite volume method), the matrix would be diagonal so the primary cost would be in the assembly. This would permit the current dimensionally split approach to be used at least temporarily. Any actual VoF algorithm is likely to need a number of steps, so the requirement of a separate assemble/solve setup is likely to remain. If limited to the current signals  assembly/solution would need to either be done in a post_stokes_solver signal which might be considered an undesirable approach especially as the computation time would be counted in the stokes solve timer, or as part of a new signal. Whichever approach is preferred for this would likely also make sense to use for the particle advection in #829.

**Comment by class4kayaker** [comment]
Closed due to major adjustment of implementation in #835.
